synapse/lib/stormtypes.py

```diff
@@ -3243,11 +3286,19 @@ async def _methSetHas(self, item):
         return item in self.valu
 
     async def _methSetAdd(self, *items):
-        [self.valu.add(i) for i in items]
+        for i in items:
+            if ismutable(i):
+                mesg = f'{await torepr(i)} is mutable and cannot be used in a set.'
+                raise s_exc.StormRuntimeError(mesg=mesg)
+            self.valu.add(i)
 
     async def _methSetAdds(self, *items):
         for item in items:
-            [self.valu.add(i) async for i in toiter(item)]
+            async for i in toiter(item):
+                if ismutable(i):
```

Comment:
could/would/should we make things like Text get collapsed to an immutable type automatically since it's probably what they mean?  Could some mutable types be singletons that we could differentiate membership by `is` comparisons?

Reply:
Wouldn't that introduce some strange behavior though? Like if I keep around both a set object and a text object, suddenly I could have this split between what's in my set versus the variable I have a direct reference to. If I then update the text object, does the set object update? If it doesn't, that seems kinda confusing to me. I would almost expect it to? But that might just be my expectations of python leaking in a bit.

Plus, we do have some easy type conversions available that are just `foo.str()` away, so I figured making magic behind the scenes happen like that wasn't a big priority.

As for singletons, I thought about the Lib* classes being singletons, butt each lib reference is less a handle and more an instance with it's own internal state that could differ from instance to instance, At least for the imported libs, maybe the built-in libs could be treated that way, but we don't really differentiate the built-ins strongly from custom libs. Aside from that, none of the others felt like they could be singletons in some arbitrary storm program.